### PR TITLE
perf(state): Cache the block hash

### LINKED ISSUE #2923: Cache computation of the block hash in state
### Summary

We re-compute the block hash several times during consensus. We should just cache this block hash computation. E.g. each `HashesTo` call recreates the block's hash.

Here is the time spent in block hashing in a node synced to osmosis mainnet for 1 hour: 
image

This is by no means the bottleneck, but is part of the sequential bottleneck of consensus and would be good to reduce, via only computing it once per block.

## Proposal

I suggest that we cache this by adding a `verifiedHash` field to the internal `Block` type.

### PR DESCRIPTION
**Pull request by ValarDragon**
Closes #2923 

Caches the block hash to ensure we only compute it once in consensus execution.

---

#### PR checklist

- [x] Tests written/updated - not sure what test I should write, any suggestions?
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

**Review by melekes — APPROVED** [review]
Thanks @ValarDragon ❤️

**Review by hvanz — APPROVED** [review]

**Review by cason — COMMENTED** [review]
Do you know why we need a mutex here?

File: types/block.go [hunk 1]
```diff
@@ -44,10 +44,11 @@ const (
 type Block struct {
 	mtx cmtsync.Mutex
 
-	Header     `json:"header"`
-	Data       `json:"data"`
-	Evidence   EvidenceData `json:"evidence"`
-	LastCommit *Commit      `json:"last_commit"`
+	verifiedHash cmtbytes.HexBytes // Verified block hash (not included in the struct hash)
```
Comment (cason) [inline types/block.go:47]: You could skip a line, to separate private from public field, just a nit.

**Comment by ValarDragon** [comment]
I don't know why the mutex is needed

**Pull request merged**